plotly/plotly/plotly.py

```diff
@@ -119,21 +154,52 @@ def iplot(figure_or_data, **plot_options):
         'extend': add additional numbers (data) to existing traces
         'append': add additional traces to existing data lists
     world_readable (default=True) -- make this figure private/public
-
+    sharing ('public' | 'private' | 'sharing') -- Toggle who can view this
+                                                  graph
+        - 'public': Anyone can view this graph. It will appear in your profile
+                    and can appear in search engines. You do not need to be
+                    logged in to Plotly to view this chart.
+        - 'private': Only you can view this plot. It will not appear in the
+                     Plotly feed, your profile, or search engines. You must be
+                     logged in to Plotly to view this graph. You can privately
+                     share this graph with other Plotly users in your online
+                     Plotly account and they will need to be logged in to
+                     view this plot.
+        - 'secret': Anyone with this secret link can view this chart. It will
+                    not appear in the Plotly feed, your profile, or search
+                    engines. If it is embedded inside a webpage or an IPython
+                    notebook, anybody who is viewing that page will be able to
+                    view the graph. You do not need to be logged in to view
+                    this plot.
     """
     if 'auto_open' not in plot_options:
         plot_options['auto_open'] = False
-    res = plot(figure_or_data, **plot_options)
-    urlsplit = res.split('/')
-    username, plot_id = urlsplit[-2][1:], urlsplit[-1]  # TODO: HACKY!
+    url = plot(figure_or_data, **plot_options)
+    urlsplit = six.moves.urllib.parse.urlparse(url)
+    file_id = urlsplit.path.split('/')[2]
+
+    if isinstance(figure_or_data, dict):
+        layout = figure_or_data.get('layout', {})
+    else:
+        layout = {}
+
+    width = layout.get('width', '100%')
+    height = layout.get('height', 525)
+    try:
+        float(width)
+    except (ValueError, TypeError):
+        pass
+    else:
+        width = str(width) + 'px'
 
-    embed_options = dict()
-    if 'width' in plot_options:
-        embed_options['width'] = plot_options['width']
-    if 'height' in plot_options:
-        embed_options['height'] = plot_options['height']
+    try:
+        float(height)
+    except (ValueError, TypeError):
+        pass
+    else:
+        height = str(height) + 'px'
 
-    return tools.embed(username, plot_id, **embed_options)
+    return tools.embed(url, file_id, width, height)
```

Comment:
I believe either (a) this should break or (b) we've made `tools.embed` backwards incompatible. The way it *used* to work was to allow either `tools.embed(url, **embed_options)` or `tools.embed(username, file_id, **embed_options)`

By calling it like `tools.embed(url, file_id, **embed_options)`, it *should* think that the url param is a username and I'd imagine it would break.